app/scripts/modules/core/src/presentation/ModalService.tsx

```diff
@@ -1,11 +1,52 @@
-import { ReactElement } from 'react';
 import * as React from 'react';
+import { Modal, ModalProps } from 'react-bootstrap';
 import * as ReactDOM from 'react-dom';
 
+import { pick, omit } from 'lodash';
+
+/** The Modal content Component will be passed these two props */
+export interface IModalComponentProps {
+  // Close modal with a result value (i.e., OK button)
+  closeModal?(result: any): void;
+
+  // Dismiss/reject modal (i.e., Cancel button)
+  dismissModal?(rejectReason: any): void;
+}
+
 /** An imperative service for showing a react component as a modal */
 export class ReactModal {
-  public static show(modal: ReactElement<any>): Promise<any> {
-    return new Promise((resolve) => {
+  /**
+   * example:
+   * const MyComponent = ({ closeModal, dismissModal }) => {
+   *   <h1>Modal Contents!</h1>
+   *   <button onClick={() => closeModal('A')}>Choice A</button>
+   *   <button onClick={() => closeModal('B')}>Choice B</button>
+   *   <button onClick={() => dismissModal('cancelled')}>Cancel</button>
+   * }
+   *
+   * ...
+   *
+   * ModalService.show<string>(MyComponent).then(result => {
+   *   this.setState({ result });
+   * });
+   *
+   * @param ModalComponent the component to be rendered inside a modal
+   * @param props to pass to the Modal and ModalComponent
+   * @returns {Promise<T>}
+   */
+  public static show<P extends IModalComponentProps, T = any>(ModalComponent: React.ComponentType<P>, props?: P): Promise<T> {
+    const modalPropKeys: [keyof ModalProps] = [
+      'onHide', 'animation', 'autoFocus', 'backdrop', 'backdropClassName', 'backdropStyle',
+      'backdropTransitionTimeout', 'bsSize', 'container', 'containerClassName', 'dialogClassName',
+      'dialogComponent', 'dialogTransitionTimeout', 'enforceFocus', 'keyboard', 'onBackdropClick',
+      'onEnter', 'onEntered', 'onEntering', 'onEscapeKeyUp', 'onExit', 'onExited', 'onExiting',
+      'onShow', 'transition',
```

Comment:
I like how this list is mostly alphabetical, but then you throw `onHide` at the beginning to keep us on our toes.

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/38a16c7ac7a23708e5bcb0b40239ccf08ac136bb/types/react-bootstrap/lib/Modal.d.ts#L10-L34

Reply:
I liked my story better